src/GafferImage/DeepSlice.cpp

```diff
@@ -0,0 +1,1047 @@
+		// I'm OK with simply discarding a volume sample when we are taking 0% of it, even if its alpha is 1,
+		// which keeps the code simpler.
+
+		int offset = sampleOffsets[i];
+		int start = prevOffset;
+		if( nearClip )
+		{
+			// Increment start to omit any samples that are before the near clip, but don't skip a point
+			// sample exactly at the near clip ( if zBack is exactly on the clip, we only skip if it's
+			// a volume sample with z < zBack )
+			while( start < offset && ( zBack[start] < nearClipDepth || ( zBack[start] == nearClipDepth && z[start] < nearClipDepth ) ) )
+			{
+				start++;
+			}
+		}
+
+		int end = offset;
+		if( farClip )
+		{
+			end = start;
+
+			// Increment end to include any samples that are strictly before the far clip. Point samples exactly
+			// at the near clip are omitted.
+			while( end < offset && z[end] < farClipDepth )
+			{
+				end++;
+			}
+		}
+
+		accumCount += end - start;
+		accumCounts[i] = accumCount;
+		inputIndices[i] = start;
+
+		// Now set the weights for what fractions of the input samples to take.
+		//
+		// Note: you'll see a bunch of cases skipping these calculations when zBack == z.
+		// That's a point sample, which we either take or don't, you can't slice a fraction
+		// of a sample with no size.
+		float firstWeight = 1.0f;
+		float lastWeight = 1.0f;
+		if( end - start == 0 )
+		{
+			// If there are no samples, no need to worry about the values of the multipliers
+		}
+		else if( nearClip && farClip && end - start == 1 && zBack[start] > z[start] )
+		{
+			// Weird special case: if there is exactly one sample, then there's the possibility
+			// that both the start and end of the sample could be clipped. We put the combined
+			// weight representing both clips into firstWeight, and leave lastWeight set to 1.0.
+
+			float usedFraction =
+				( std::min( zBack[start], farClipDepth ) - std::max( z[start], nearClipDepth ) ) /
+				( zBack[start] - z[start] );
+
+			firstWeight = sampleMultiplier( a ? a[start] : 0.0f, usedFraction );
+		}
+		else
+		{
+			if( nearClip && zBack[start] > z[start] )
+			{
+				float usedFraction =
+					( zBack[start] - std::max( z[start], nearClipDepth ) ) /
+					( zBack[start] - z[start] );
+
+				firstWeight = sampleMultiplier( a ? a[start] : 0.0f, usedFraction );
+			}
+
+			if( farClip && zBack[end - 1] > z[end - 1] )
+			{
+				float usedFraction =
+					( std::min( zBack[end - 1], farClipDepth ) - z[end - 1] ) /
+					( zBack[end - 1] - z[end - 1] );
+
+				lastWeight = sampleMultiplier( a ? a[end - 1] : 0.0f, usedFraction );
+			}
+		}
+
+		if( firstWeights )
+		{
+			firstWeights[i] = firstWeight;
+		}
+
+		if( lastWeights )
+		{
+			lastWeights[i] = lastWeight;
+		}
+
+		prevOffset = offset;
+	}
+
+	// Fill the result CompoundObject
+
+	CompoundObjectPtr result = new CompoundObject;
+	result->members()[ g_accumCountsName ] = std::move( accumCountsData );
```

Comment:
Hmm, I've been trying to get into the habit of not incurring extra reference counts. It's true that the impact of that is probably not measurable in this case, but it seems like this is probably the better habit to be in? When you ask if it's worth it, what cost are we asking if it's worth? Typing 11 more characters? I guess my proposed principle is more: when initializing the members of a CompoundObject, you should only do a double-increment of the reference counts if there is some reason you need to do that?

Reply:
I guess my thought is that "constructing a CompoundObject or CompoundData" is a pretty common pattern, and we probably should be doing this everywhere we use this pattern, since some of them probably are a lot closer to being performance sensitive. And if we start doing this everywhere we construct a CompoundObject, then doing it here would potentially have low to negative mental cost - we should be asking "why not" any time we see this sort of CompoundObject construction without std::move.

> we don't do this elsewhere

My assumption was that the only reason we're not doing this generally was just because it's only fairly recently that we've been able to use std::move.

Glancing at the codebase, I can see several places where it would certainly make sense to me:
https://github.com/GafferHQ/gaffer/blob/main/src/Gaffer/Spreadsheet.cpp#L1172
https://github.com/GafferHQ/gaffer/blob/main/src/GafferScene/Options.cpp#L124

Those are more plausibly performance sensitive than DeepSlice, though admittedly I can't see any smoking gun where the difference in performance would be really significant.

A lot of this does come down purely to perspective - you see it as a mental cost to notice an std::move, I kinda feel like we ought to be more more aware of the reference counting that is happening when a smart pointer assignment happens ... if you see an std::move, then less is actually going to happen, and you should have less to think about.

> having to check that we don't use the moved-from objects afterwards

This is another one where it almost feels to me more like reducing mental load than increasing it - an std::move clearly documents intent that the variable will not be used afterwards ( backed by a nice handy segfault if that documentation is wrong )